server/src/test/java/org/elasticsearch/search/SearchHitsTests.java

```diff
@@ -21,88 +21,126 @@
+                    instance.getTotalHits(), instance.getMaxScore());
+            case 1:
+                final TotalHits totalHits;
+                if (instance.getTotalHits() == null) {
+                    totalHits = randomTotalHits();
+                } else {
+                    totalHits = null;
+                }
+                return new SearchHits(instance.getHits(), totalHits, instance.getMaxScore());
+            case 2:
+                final float maxScore;
+                if (Float.isNaN(instance.getMaxScore())) {
+                    maxScore = randomFloat();
+                } else {
+                    maxScore = Float.NaN;
+                }
+                return new SearchHits(instance.getHits(), instance.getTotalHits(), maxScore);
+            default:
+                throw new UnsupportedOperationException();
+        }
+    }
+
+    @Override
+    protected Predicate<String> getRandomFieldsExcludeFilter() {
+        return path -> (path.isEmpty() ||
+            path.contains("inner_hits") || path.contains("highlight") || path.contains("fields") || path.contains("_source"));
+    }
+
+    @Override
+    protected String[] getShuffleFieldsExceptions() {
+        return new String[] {"_source"};
     }
 
     @Override
     protected SearchHits createBlankInstance() {
-        return new SearchHits(new SearchHit[0], new TotalHits(0, TotalHits.Relation.EQUAL_TO), Float.NaN);
+        return new SearchHits();
     }
 
     @Override
     protected SearchHits createTestInstance() {
-        return createTestItem();
+        return createTestItem(randomFrom(XContentType.values()), true, true);
     }
 
-    public void testFromXContent() throws IOException {
-        SearchHits searchHits = createTestItem();
-        XContentType xcontentType = randomFrom(XContentType.values());
-        boolean humanReadable = randomBoolean();
-        BytesReference originalBytes = toShuffledXContent(searchHits, xcontentType, ToXContent.EMPTY_PARAMS, humanReadable);
-        SearchHits parsed;
-        try (XContentParser parser = createParser(xcontentType.xContent(), originalBytes)) {
-            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
-            assertEquals(XContentParser.Token.FIELD_NAME, parser.nextToken());
-            assertEquals(SearchHits.Fields.HITS, parser.currentName());
-            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
-            parsed = SearchHits.fromXContent(parser);
-            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
-            assertEquals(XContentParser.Token.END_OBJECT, parser.nextToken());
-            assertNull(parser.nextToken());
-        }
-        assertToXContentEquivalent(originalBytes, toXContent(parsed, xcontentType, humanReadable), xcontentType);
+    @Override
+    protected SearchHits createXContextTestInstance(XContentType xContentType) {
+        return createTestItem(xContentType,true, false);
```

Comment:
I think that we never exercise the codepath that prints out (and parses back) the shard info when explanation is set. if it becomes complicated to exercise that as part of our randomized tests (by setting the target in some cases), shall we have a small specialized test method to cover that bit? We could also do that in SearchHitTests for a single hit rather than here.

Reply:
I added a test in https://github.com/elastic/elasticsearch/pull/36313/commits/d6e724a6520bab25597bb6859c139dda674f85e3